boltdesignsystem/bolt
# ES Modules: Base Component Function to Class Migration + Dependency Cleanup

## Jira
N/A

## Summary
Migrates all existing components in Bolt to use a Class-based Base Component (vs the function-based Base Component in use currently). This is one of THE most important pieces to the broader ES Module-related work in https://github.com/boltdesignsystem/bolt/pull/1579.

As an added bonus, not only will Class-based components further improve Typescript / JSDoc support in Bolt, they are also a central part of the performance improvements this ES Module work should result in. 

## Details
- Lit-html and Preact-based base classes in use are no longer function-based (note how there aren't parentheses getting used in almost all of the component Class definitions)
- Because of this, the `@define` decorator is no longer in use (not supported with the new plumbing) and all component instances switch over to use the new `@bolt/element` `@customElement('component-name')` syntax instead
- With the `customElement` decorator in use, I also cleaned up and consolidated a handful of common dependencies used, namely lit-html based dependencies like `html` or `unsafeHTML`
- Additionally, because we're now using Class-based component definitions, all of our custom decorators created will no longer work — work that's taken care of in https://github.com/boltdesignsystem/bolt/pull/1586

## How to Test
In a nutshell, you can't, at least not yet. 

In order to split up the code changes from https://github.com/boltdesignsystem/bolt/pull/1579 into more sensible chunks to review, these component updates require virtually [every update](https://github.com/boltdesignsystem/bolt/pull/1579#user-content-related-prs) listed out in these related PRs required to ship section in order to compile and work as expected.
